module/VuFind/src/VuFind/ILS/Driver/Folio.php

```diff
@@ -458,45 +471,56 @@ public function getHolding($bibId, array $patron = null, array $options = [])
      */
     public function patronLogin($username, $password)
     {
-        // Get user id
-        $query = ['query' => 'username == ' . $username];
-        $response = $this->makeRequest('GET', '/users', $query);
+        if ($this->config['User']['okapi_login'] ?? false) {
+            $tenant = $this->config['API']['tenant'];
+            $credentials = compact('tenant', 'username', 'password');
+            // Get token
+            try {
+                $response = $this->makeRequest(
+                    'POST',
+                    '/authn/login',
+                    json_encode($credentials)
+                );
+                $debugMsg = 'User logged in. User: ' . $username . '.';
+                // We've authenticated the user with Okapi, but we only have their
+                // username; set up a query to retrieve full info below.
+                $query = 'username == ' . $username;
+                // Replace admin with user as tenant if configured to do so:
+                if ($this->config['User']['use_user_token'] ?? false) {
+                    $this->token = $response->getHeaders()->get('X-Okapi-Token')
+                        ->getFieldValue();
+                    $debugMsg .= ' Token: ' . substr($this->token, 0, 30) . '...';
+                }
+                $this->debug($debugMsg);
+            } catch (Exception $e) {
+                return null;
+            }
+        } else {
+            // Construct user query using barcode, username, etc.
+            $usernameField = $this->config['User']['username_field'] ?? 'username';
+            $passwordField = $this->config['User']['password_field'] ?? 'password';
+            $query = $usernameField . ' == ' . $this->escapeCql($username);
+            if (!empty($passwordField)) {
```

Comment:
I added back the optional passwordField setting, in case you want to do a two-factor verification using LDAP attributes in both cat_username and cat_password. Is there any chance you'd be able to test if this actually works?

Reply:
From what I can tell (at least the way we have LDAP set up) this cannot be done.  When I try to call /auth/login with my own LDAP userid/password (which works in our FOLIO UX) ...the message I get back is:  "No credentials match that login" @ihardy - Do you know if this is supposed to work like this...or do you think we may have something misconfigured in FOLIO? thank you!

Reply:
Just to be clear, I'm not suggesting logging in to FOLIO using your LDAP credentials, but rather setting up VuFind's LDAP login so that it sets both cat_username and cat_password with different LDAP attributes that can be matched in FOLIO (for example, barcode and last name), and then setting the usernameField/passwordField in FOLIO.ini to properly map the attribute values into a FOLIO CQL query. Does that make any sense?

Reply:
ah...sorry for the confusion.  I'm going to have to take a minute to get my mind wrapped around this.  :)